server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java

```diff
@@ -0,0 +1,234 @@
+                        nodesToBeDecommissioned.toString()
+                    )
+                );
+            }
+        }, allDecommissionedNodesRemovedPredicate);
+    }
+
+    public void updateMetadataWithDecommissionStatus(DecommissionStatus decommissionStatus, ActionListener<Void> listener) {
+        clusterService.submitStateUpdateTask(decommissionStatus.status(), new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                Metadata metadata = currentState.metadata();
+                DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.custom(DecommissionAttributeMetadata.TYPE);
+                assert decommissionAttributeMetadata != null && decommissionAttributeMetadata.decommissionAttribute() != null;
+                assert assertStatusTransitionOrFailed(decommissionAttributeMetadata.status(), decommissionStatus);
+                Metadata.Builder mdBuilder = Metadata.builder(metadata);
+                DecommissionAttributeMetadata newMetadata = decommissionAttributeMetadata.withUpdatedStatus(decommissionStatus);
+                mdBuilder.putCustom(DecommissionAttributeMetadata.TYPE, newMetadata);
+                return ClusterState.builder(currentState).metadata(mdBuilder).build();
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                DecommissionAttributeMetadata decommissionAttributeMetadata = newState.metadata()
+                    .custom(DecommissionAttributeMetadata.TYPE);
+                assert decommissionAttributeMetadata.status().equals(decommissionStatus);
+                listener.onResponse(null);
+            }
+        });
+    }
+
+    private static boolean assertStatusTransitionOrFailed(DecommissionStatus oldStatus, DecommissionStatus newStatus) {
+        switch (newStatus) {
+            case DECOMMISSION_INIT:
+                // if the new status is INIT, then the old status cannot be anything but FAILED
+                return oldStatus.equals(DecommissionStatus.DECOMMISSION_FAILED);
+            case DECOMMISSION_IN_PROGRESS:
+                // if the new status is IN_PROGRESS, the old status has to be INIT
+                return oldStatus.equals(DecommissionStatus.DECOMMISSION_INIT);
+            case DECOMMISSION_SUCCESSFUL:
+                // if the new status is SUCCESSFUL, the old status has to be IN_PROGRESS
+                return oldStatus.equals(DecommissionStatus.DECOMMISSION_IN_PROGRESS);
+            default:
+                // if the new status is FAILED, we don't need to assert for previous state
+                return true;
```

Comment:
Lets avoid the default and put explicit checks

Reply:
Okay